parcels/spatialhash.py

```diff
@@ -0,0 +1,326 @@
+# from __future__ import annotations
+
+# from typing import Literal
+
+# import numpy as np
+# import uxarray as ux
+# from uxarray.grid.coordinates import _lonlat_rad_to_xyz
+
+# from parcels.field import FieldOutOfBoundError  # Adjust import as necessary
+# from parcels.xgrid import _search_1d_array
+
+# from .basegrid import BaseGrid
+import numpy as np
+
+
+class SpatialHash:
+    """Custom data structure that is used for performing grid searches using Spatial Hashing. This class constructs an overlying
+    uniformly spaced rectilinear grid, called the "hash grid" on top parcels.xgrid.XGrid. It is particularly useful for grid searching
+    on curvilinear grids. Faces in the Xgrid are related to the cells in the hash grid by determining the hash cells the bounding box
+    of the unstructured face cells overlap with.
+
+    Parameters
+    ----------
+    grid : parcels.xgrid.XGrid
+        Source grid used to construct the hash grid and hash table
+    global_grid : bool, default=False
+        If true, the hash grid is constructed using the domain [-pi,pi] x [-pi,pi]
+    reconstruct : bool, default=False
+        If true, reconstructs the spatial hash
+
+    Note
+    ----
+    Does not currently support queries on periodic elements.
+    """
+
+    def __init__(
+        self,
+        grid,
+        global_grid=False,
+        reconstruct=False,
+    ):
+        # TODO : Enforce grid to be an instance of parcels.xgrid.XGrid and curvilinear.
+        self._source_grid = grid
+        self.reconstruct = reconstruct
+
+        # Hash grid size
+        self._dh = self._hash_cell_size()
+
+        if global_grid:
+            # Set the hash grid to be a global grid
+            self._xmin = -180.0
+            self._ymin = -90.0
+            self._xmax = 180.0
+            self._ymax = 90.0
+        else:
+            # Lower left corner of the hash grid
+            lon_min = self._source_grid.lon.min()
+            lat_min = self._source_grid.lat.min()
+            lon_max = self._source_grid.lon.max()
+            lat_max = self._source_grid.lat.max()
+
+            self._xmin = lon_min - self._dh
+            self._ymin = lat_min - self._dh
+            self._xmax = lon_max + self._dh
+            self._ymax = lat_max + self._dh
```

Comment:
Not sure the motivation for the `global_grid` attribute in this PR - don't we always have access to the lons and lats? So we can just access this from the grid object? This flag might be error prone (e.g., if dataset is provided with lon [0,360])

Reply:
I've removed the `global_grid` parameter in https://github.com/OceanParcels/Parcels/pull/2132/commits/27b16eeeafce6fa21c38354ac2b565622cfe5413